recipes-extended/quark/files/add-POST-and-v4v-support.patch

```diff
@@ -0,0 +1,319 @@
++	wait_fd_ready(fd, true, true);
++	ret = resp_file(fd, RELPATH(realtarget), r, &st, mime, lower, upper);
++	wait_fd_ready(fd, true, false);
++
++	return ret;
+ }
+diff --git a/http.h b/http.h
+index cd1ba22..f0ea49e 100644
+--- a/http.h
++++ b/http.h
+@@ -5,6 +5,7 @@
+ #include <limits.h>
+ 
+ #define HEADER_MAX 4096
++#define PAYLOAD_MAX 8192
+ #define FIELD_MAX 200
+ 
+ enum req_field {
+@@ -19,6 +20,7 @@ extern const char *req_field_str[];
+ enum req_method {
+ 	M_GET,
+ 	M_HEAD,
++	M_POST,
+ 	NUM_REQ_METHODS,
+ };
+ 
+@@ -28,6 +30,7 @@ struct request {
+ 	enum req_method method;
+ 	char target[PATH_MAX];
+ 	char field[NUM_REQ_FIELDS][FIELD_MAX];
++	char payload[PAYLOAD_MAX];
+ };
+ 
+ enum status {
+diff --git a/main.c b/main.c
+index e9d0a06..a3eebf6 100644
+--- a/main.c
++++ b/main.c
+@@ -13,6 +13,7 @@
+ #include <stdio.h>
+ #include <stdlib.h>
+ #include <string.h>
++#include <syslog.h>
+ #include <time.h>
+ #include <unistd.h>
+ 
+@@ -31,14 +32,17 @@ serve(int infd, struct sockaddr_storage *in_sa)
+ 	char inaddr[INET6_ADDRSTRLEN /* > INET_ADDRSTRLEN */];
+ 	char tstmp[21];
+ 
+-	/* set connection timeout */
+-	if (sock_set_timeout(infd, 30)) {
+-		goto cleanup;
+-	}
+-
+ 	/* handle request */
+-	if (!(status = http_get_request(infd, &r))) {
+-		status = http_send_response(infd, &r);
++	status = http_get_request(infd, &r);
++
++	if (!status) {
++		if (r.method == M_POST) {
++			/* handle POST */
++			syslog(LOG_INFO, "[quark] %s", r.payload);
++		} else {
++			/* handle GET */
++			status = http_send_response(infd, &r);
++		}
+ 	}
+ 
+ 	/* write output to log */
+@@ -228,18 +232,6 @@ main(int argc, char *argv[])
+ 		die("setrlimit RLIMIT_NPROC:");
+ 	}
+ 
+-	/* validate user and group */
+-	errno = 0;
+-	if (user && !(pwd = getpwnam(user))) {
+-		die("getpwnam '%s': %s", user, errno ? strerror(errno) :
+-		    "Entry not found");
+-	}
+-	errno = 0;
+-	if (group && !(grp = getgrnam(group))) {
+-		die("getgrnam '%s': %s", group, errno ? strerror(errno) :
+-		    "Entry not found");
+-	}
+-
+ 	/* Open a new process group */
+ 	setpgid(0,0);
+ 
+@@ -270,23 +262,6 @@ main(int argc, char *argv[])
+ 			die("chroot .:");
+ 		}
+ 
+-		/* drop root */
+-		if (grp && setgroups(1, &(grp->gr_gid)) < 0) {
+-			die("setgroups:");
+-		}
+-		if (grp && setgid(grp->gr_gid) < 0) {
+-			die("setgid:");
+-		}
+-		if (pwd && setuid(pwd->pw_uid) < 0) {
+-			die("setuid:");
+-		}
+-		if (getuid() == 0) {
+-			die("Won't run as root user", argv0);
+-		}
+-		if (getgid() == 0) {
+-			die("Won't run as root group", argv0);
+-		}
+-
+ 		/* accept incoming connections */
+ 		while (1) {
+ 			in_sa_len = sizeof(in_sa);
+diff --git a/resp.c b/resp.c
+index 3075c28..ee12606 100644
+--- a/resp.c
++++ b/resp.c
+@@ -141,6 +141,7 @@ resp_file(int fd, char *name, struct request *r, struct stat *st, char *mime,
+ 	            "Date: %s\r\n"
+ 	            "Connection: close\r\n"
+ 	            "Last-Modified: %s\r\n"
++	            "Expires: Wed, 1 Jan 2030 00:00:00 GMT\r\n"
```

Comment:
The main reason this is here is because UID sent a similar line in its response header (see https://github.com/OpenXT/uid/blob/85bdfe033454efc8016362c466599b0658d91c79/file_handler.ml#L82). It's supposed to make use of HTTP caching, but I'm not even really sure that the client is taking advantage of that. i can remove it if desired.

Reply:
UID does it on images only, this is probably to not reload the entire UI and get things a little faster, the browser likely handles the caching.
While the images in the UI page won't change, could this trigger some fun on other resources?

Also `quark` already has helpers to handle timestamps, this could be now + 10years, or something like that.

Reply:
Ah good catch I didn't realize it was just for images. I can do something similar using the helpers.